enricher/govulncheck/source/govulncheck.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/google/osv-scalibr/log"
+	"github.com/google/osv-scalibr/plugin"
+	"golang.org/x/vuln/scan"
+)
+
+const (
+	// Name is the unique name of this enricher.
+	Name = "reachability/govulncheck/source"
+)
+
+var ErrNoGoToolchain = errors.New("no Go toolchain found")
+
+// Enricher is an enricher that runs govulncheck on Go source code.
+type Enricher struct{}
+
+// Name returns the name of the enricher.
+func (e *Enricher) Name() string {
+	return Name
+}
+
+// Version returns the version of the enricher.
+func (e *Enricher) Version() int {
+	return 0
+}
+
+// Requirements returns the requirements of the enricher.
+func (e *Enricher) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{
+		Network:  plugin.NetworkOnline,
+		DirectFS: true,
+	}
+}
+
+// RequiredPlugins returns the names of the plugins required by this enricher.
+func (e *Enricher) RequiredPlugins() []string {
+	return []string{gomod.Name}
+}
+
+// NewEnricher creates the enricher
+func NewEnricher() Enricher {
+	return Enricher{}
+}
+
+// Enrich runs govulncheck on the Go modules in the inventory.
+func (e *Enricher) Enrich(ctx context.Context, input *enricher.ScanInput, inv *inventory.Inventory) error {
+	cmd := exec.CommandContext(ctx, "go", "version")
+	_, err := cmd.Output()
+	if err != nil {
+		return ErrNoGoToolchain
+	}
+
+	goModVersions := make(map[string]string)
+	for _, pkg := range inv.Packages {
+		if !slices.Contains(pkg.Plugins, gomod.Name) {
+			continue
+		}
+		if pkg.Name == "stdlib" {
+			for _, l := range pkg.Locations {
+				if goModVersions[l] != "" {
+					continue
+				}
+
+				// Set GOVERSION to the Go version in go.mod.
+				goModVersions[l] = pkg.Version
+
+				continue
+			}
+		}
+	}
+
+	for goModLocation, goVersion := range goModVersions {
+		modDir := filepath.Dir(goModLocation)
+		absModDir := filepath.Join(input.ScanRoot.Path, modDir)
+		findings, err := e.runGovulncheck(ctx, absModDir, goVersion)
+		if err != nil {
+			log.Errorf("govulncheck on %s: %v", modDir, err)
+			continue
+		}
+
+		if len(findings) == 0 {
+			continue
+		}
+
+		e.addSignals(inv, findings)
+	}
+
+	return nil
+}
+
+func (e *Enricher) addSignals(inv *inventory.Inventory, idToFindings map[string][]*Finding) {
+	for _, pv := range inv.PackageVulns {
+		findings, exist := idToFindings[pv.Vulnerability.Id]
+		// Skip if no findings for this package vulnerability ID
+		if !exist {
+			continue
+		}
+
+		isReachable := false
+		for _, f := range findings {
+			if len(f.Trace) > 0 && f.Trace[0].Function != "" {
+				isReachable = true
+				break
+			}
+		}
+
+		if !isReachable {
+			pv.ExploitabilitySignals = append(pv.ExploitabilitySignals, &vex.FindingExploitabilitySignal{
+				Plugin:        Name,
+				Justification: vex.VulnerableCodeNotInExecutePath,
+			})
+		}
+	}
+}
+
+func (e *Enricher) runGovulncheck(ctx context.Context, absModDir string, goVersion string) (map[string][]*Finding, error) {
+	cmd := scan.Command(ctx, "-C", absModDir, "-json", "./...")
```

Comment:
Does the govulncheck library support source scanning? If so then we could import the library like for the [binary detector](https://github.com/google/osv-scalibr/blob/ed0917ecede176668e40a3df292ecf8b06fdbd7f/detector/govulncheck/binary/binary.go#L130) instead of relkying on the presence of an external command